### ISSUE TITLE
JIT: Improper zero-trip checking in loop cloning/unrolling

### ISSUE THREAD

**Issue by jakobbotsch** [dotnet/runtime#96623]
The following program hits an AV at runtime. It should throw `IndexOutOfRangeException`.
```csharp
public static void Main()
{
    Foo(new int[15]);
}

private static int Foo(int[] arr)
{
    int limit = GetLimit();
    int sum = 0;
    int i = 0;
    int j = 0;

    int x;

    if (Environment.TickCount < 0)
        x = 42;
    else
        x = 42;

    if (x == 42)
    {
        i = int.MaxValue;
        goto TestInner;
    }

    Console.WriteLine("Unreachable");
    goto TestOuter;

OuterStart:;
    Console.WriteLine("Outer");
    goto TestInner;

InnerStart:;
    sum += arr[i];

TestInner:;
    j++;
    if (j < 30)
        goto InnerStart;

TestOuter:;
    i++;
    if (i < limit)
        goto OuterStart;

    return sum;
}

[MethodImpl(MethodImplOptions.NoInlining)]
private static int GetLimit() => 10;
```

The problem is the JIT assumes that the entry is protected by a zero-trip test whenever we have performed loop inversion. However, the zero-trip test created by loop inversion can be unreachable, like in this case. The flow graph looks like [this](https://edotor.net/?engine=dot?engine=dot#digraph%20FlowGraph%20%7B%0A%0A%20%20%20%20graph%20%5Blabel%20%3D%20%22Foo%5Cnafter%5CnInvert%20loops%22%5D%3B%0A%20%20%20%20node%20%5Bshape%20%3D%20%22Box%22%5D%3B%0A%20%20%20%20BB01%20%5Blabel%20%3D%20%22BB01%22%2C%20shape%20%3D%20%22house%22%5D%3B%0A%20%20%20%20BB02%20%5Blabel%20%3D%20%22BB02%22%5D%3B%0A%20%20%20%20BB03%20%5Blabel%20%3D%20%22BB03%22%5D%3B%0A%20%20%20%20BB09%20%5Blabel%20%3D%20%22BB09%5CnV03%20%3E%3D%20V01%22%2C%20shape%20%3D%20%22note%22%5D%3B%0A%20%20%20%20BB04%20%5Blabel%20%3D%20%22BB04%22%5D%3B%0A%20%20%20%20BB10%20%5Blabel%20%3D%20%22BB10%5CnV04%20%3E%3D%2030%22%2C%20shape%20%3D%20%22note%22%5D%3B%0A%20%20%20%20BB05%20%5Blabel%20%3D%20%22BB05%22%5D%3B%0A%20%20%20%20BB06%20%5Blabel%20%3D%20%22BB06%5CnV04%20%3C%2030%22%5D%3B%0A%20%20%20%20BB07%20%5Blabel%20%3D%20%22BB07%5CnV03%20%3C%20V01%22%5D%3B%0A%20%20%20%20BB08%20%5Blabel%20%3D%20%22BB08%22%2C%20shape%20%3D%20%22invhouse%22%5D%3B%0A%20%20%20%20BB01%20-%3E%20BB02%20%5Bcolor%3Dblue%2C%20weight%3D20%5D%3B%0A%20%20%20%20BB03%20-%3E%20BB09%20%5Bcolor%3Dblue%2C%20weight%3D20%5D%3B%0A%20%20%20%20BB07%20-%3E%20BB04%20%5Bcolor%3Dgreen%5D%3B%0A%20%20%20%20BB09%20-%3E%20BB04%20%5Bcolor%3Dblue%2C%20weight%3D20%5D%3B%0A%20%20%20%20BB02%20-%3E%20BB10%20%5B%5D%3B%0A%20%20%20%20BB04%20-%3E%20BB10%20%5Bcolor%3Dblue%2C%20weight%3D20%5D%3B%0A%20%20%20%20BB06%20-%3E%20BB05%20%5Bcolor%3Dgreen%5D%3B%0A%20%20%20%20BB10%20-%3E%20BB05%20%5Bcolor%3Dblue%2C%20weight%3D20%5D%3B%0A%20%20%20%20BB05%20-%3E%20BB06%20%5Bcolor%3Dblue%2C%20weight%3D20%5D%3B%0A%20%20%20%20BB06%20-%3E%20BB07%20%5Bcolor%3Dblue%2C%20weight%3D20%5D%3B%0A%20%20%20%20BB10%20-%3E%20BB07%20%5B%5D%3B%0A%20%20%20%20BB07%20-%3E%20BB08%20%5Bcolor%3Dblue%2C%20weight%3D20%5D%3B%0A%20%20%20%20BB09%20-%3E%20BB08%20%5B%5D%3B%0A%20%20%20%20BB01%20-%3E%20BB02%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB02%20-%3E%20BB03%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB03%20-%3E%20BB09%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB09%20-%3E%20BB04%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB04%20-%3E%20BB10%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB10%20-%3E%20BB05%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB05%20-%3E%20BB06%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB06%20-%3E%20BB07%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB07%20-%3E%20BB08%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%7D%0A) after loop inversion, where the zero-trip test in `BB09` is unreachable. This in turn makes the checking done by loop cloning insufficient to prove that the bounds check for `arr[i]` can be removed.

Sadly OSR compilations seem to make good use of this unsoundness, so fixing this bug will mean we stop loop cloning in a bunch of OSR cases.

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib @AndyAyersMS. Likely the only remedy to get the loop cloning back for OSR cases will be to insert the entry branch later in the JIT, e.g. right before lowering, so that we retain all the context of surrounding IR during optimizations. I suppose it might also be possible to place the patchpoints in a way that the zero-trip test happens at the beginning after an OSR transition, instead of at the end. Otherwise we simply do not have the context during loop cloning in these OSR cases to be able to prove that the loop invariant is true during the first iteration.

https://github.com/dotnet/performance/blob/49d4de8e6c45036b902be1eea9bb90b63b7657fb/src/benchmarks/micro/runtime/Benchstones/BenchI/BubbleSort2.cs#L16-L27
is one example where the OSR case is doing 'unproven' loop cloning right now; we eliminate the bounds check on `x[i]` without knowing that `i <= limit1` in the first iteration.

**Comment by jakobbotsch** [comment]
> I was trying to repro the failure on 8.0 but couldn't. Is this new behavior in 9.0? Does it require low OSR thresholds? I'll have to look more closely at how the flow graph evolves.
> 
> If we decide we need to defer changing flow until even later we need to be careful that we haven't done any **other** optimizations that would then become invalid. I think the list starts to grow.

The example program fails with `DOTNET_TieredCompilation=0` (no OSR necessary), but when fixing it I've noticed that it is going to result in a number of OSR regressions. 

It definitely seems hard to defer the flow change more, we would also need to make sure we checked the loop cloning conditions before jumping into one of the two versions of cloned loops. That doesn't really seem feasible. So it seems like the only thing we can do is teach loop cloning to handle the extra cases produced by OSR, for example by duplicating the zero-trip test itself as part of the choice conditions.
